lib/rubygems/package.rb

```diff
@@ -708,7 +708,11 @@ def verify_files(gem)
 
   def verify_gz(entry) # :nodoc:
     Zlib::GzipReader.wrap entry do |gzio|
-      gzio.read 16_384 until gzio.eof? # gzip checksum verification
+      buf = String.new(capacity: 16_384, encoding: Encoding::BINARY)
+      begin
+        gzio.readpartial(16_384, buf) until gzio.eof? # gzip checksum verification
```

Comment:
`gzio` read partial returns arbitrarily different length strings on some gzips. Check the pathological gem that I mentioned in the zlib bug. It extracts validly when using read but not readpartial. I don't know how common this problem is.